### PR TITLE
#9091: Create user deployment UI

### PR DESCRIPTION
**Pull request by fungairino**
## What does this PR do?

- Part of #9091
- This set of changes includes the UI for the new personal synchronize settings section in the mod activation page
- Includes feature flag gating
- Includes validation to display an error if the user attempts to activate with a local integration selected
- No actual functionality to create the personal deployment, just focusing on the UI and that the form state is correct in this change

## Discussion

- wizardStateFactory is getting a bit unwieldy with this change adding two new prop arguments. Is now a good time to consider refactoring? 

## Demo

Screenshot 2024-09-11 at 10 13 36 AM
Screenshot 2024-09-11 at 10 17 08 AM

## Future Work

- Actually making the API request to the backend to create the personal deployment when this option is true.
- UI for the sidebar mod panel activation (not done in this PR)

## Checklist

_Leave all that are relevant and check off as completed_

- [ ] This PR requires a security review
- [ ] This PR introduces a new library: double check it's MIT/Apache2/permissively licensed
- [ ] This PR requires a node/npm version update: let the team know on #engineering
- [ ] This PR requires a documentation change (link to old docs)
- [ ] This PR requires a tutorial update (link to old tutorials)
- [X] This PR requires a feature flag
- [ ] This PR requires a environment variable change
- [X] Added jest or playwright tests and/or storybook stories

For more information on our expectations for the PR process, see the
[code review principles doc](https://www.notion.so/pixiebrix/Code-Review-Principles-1ce7276b82a84d2a995d55ad85e1310d?pvs=4)

### CODE REVIEW

File: src/sidebar/activateMod/ActivateMultipleModsPanel.tsx [hunk 1]
```diff
@@ -95,6 +97,7 @@ const AutoActivatePanel: React.FC<{ mods: RequiredModDefinition[] }> = ({
           );
 
           const wizard = wizardStateFactory({
+            flagOn,
```
Comment (fungairino) [inline src/sidebar/activateMod/ActivateMultipleModsPanel.tsx:100]: I need to include authOptions here as well

File: src/activation/useActivateModWizard.ts [hunk 2]
```diff
@@ -209,6 +245,10 @@ function useActivateModWizard(
     modDefinition.options?.schema,
   );
 
+  const { flagOn } = useFlags();
+
+  const { data: authOptions } = useAuthOptions();
```
Comment (fungairino) [inline src/activation/useActivateModWizard.ts:250]: I should wrap this in a default empty array just to be safe (as is done in some other places

File: src/hooks/auth.test.ts [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+import { useGetIntegrationAuthsQuery } from "@/data/service/api";
+import { renderHook } from "@testing-library/react-hooks";
+import { type IntegrationConfig } from "@/integrations/integrationTypes";
+import { type UUID } from "@/types/stringTypes";
+import { type RegistryId } from "@/types/registryTypes";
+
+jest.mock("@/integrations/util/readRawConfigurations");
+jest.mock("@/data/service/api");
+
+describe("useAuthOptions", () => {
```
Comment (fungairino) [inline src/hooks/auth.test.ts:30]: No existing coverage for this hook, so I added it in this PR.

File: src/activation/SynchronizeBody.tsx [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import React from "react";
+import BooleanWidget from "@/components/fields/schemaFields/widgets/BooleanWidget";
+import ConnectedFieldTemplate from "@/components/form/ConnectedFieldTemplate";
+
+const SynchronizeBody: React.FunctionComponent = () => (
```
Comment (grahamlangford) [inline src/activation/SynchronizeBody.tsx:22]: I'm not fond of the name here. I believe something like `SyncPersonalDeploymentTemplate` would be more descriptive?

File: src/activation/useActivateModWizard.test.tsx [hunk 5]
```diff
@@ -16,27 +16,36 @@
...
 } from "@/auth/featureFlags";
 import { sharingDefinitionFactory } from "@/testUtils/factories/registryFactories";
 import { BusinessError } from "@/errors/businessErrors";
+import { type RegistryId } from "@/types/registryTypes";
 
 jest.mock("@/components/integrations/AuthWidget", () => {});
 jest.mock("react-redux");
 jest.mock("@/data/service/api", () => ({
   useGetFeatureFlagsQuery: jest.fn(() => valueToAsyncState([])),
+  useGetIntegrationAuthsQuery: jest.fn(() => valueToAsyncState([])),
```
Comment (grahamlangford) [inline src/activation/useActivateModWizard.test.tsx:48]: It would probably be better to use axios mock adapter here instead of mocking the api layer

Comment (fungairino) [reply src/activation/useActivateModWizard.test.tsx:48]: How strongly do you feel about this? I tend to agree that given the choice it's better to mock at the api client layer, but this test file already was mocking the rtk hook methods, and I tend to try to follow local conventions if possible - or change the whole file to use a better convention.

Comment (fungairino) [reply src/activation/useActivateModWizard.test.tsx:48]: Generally this comment has been made on a lot of unit test file PR changes so I'd like to align on a general rule going forward on when we should/should not mock the hooks vs the apiClient http call and possibly define a eslint rule for this.

File: src/activation/SynchronizeBody.tsx (hunk 4 above)
Comment (fungairino) [reply src/activation/SynchronizeBody.tsx:22]: I decided on `PersonalDeploymentField` if that sounds good to you

File: src/activation/useActivateModWizard.test.tsx (hunk 5 above)
Comment (fungairino) [reply src/activation/useActivateModWizard.test.tsx:48]: Would you be ok with keeping the data/service/api mock here in this PR and tackle a general eslint rule / refactoring for this pattern captured in a separate dev ex issue? It's non-trivial to switch it over in this test.

Comment (grahamlangford) [reply src/activation/useActivateModWizard.test.tsx:48]: I'm ok with keeping it as-is for now. For context, mocking the hook _used_ to be the norm. We've been moving towards mocking axios instead. Similar to how we used to use snapshots more

File: src/extensionConsole/pages/activateMod/ActivateModPage.test.tsx [hunk 6]
```diff
@@ -168,6 +165,51 @@ describe("ActivateModDefinitionPage", () => {
...
+
+    setupMod(modDefinition);
+
+    const { container } = render(<ActivateModDefinitionPageWrapper />);
+
+    await waitForEffect();
+
+    await waitFor(() => {
+      expect(screen.getByText("Synchronize Settings")).toBeInTheDocument();
+    });
```
Comment (grahamlangford) [inline src/extensionConsole/pages/activateMod/ActivateModPage.test.tsx:199]:
```suggestion
    await expect(screen.findByText("Synchronize Settings")).resolves.toBeInTheDocument();
    });
```

File: src/extensionConsole/pages/activateMod/ActivateModPage.test.tsx [hunk 7]
```diff
@@ -168,6 +165,51 @@ describe("ActivateModDefinitionPage", () => {
...
+
+    expect(activateModCallbackMock).toHaveBeenCalledWith(
+      {
+        modComponents: { "0": true },
+        optionsArgs: {},
+        integrationDependencies: [],
+        personalDeployment: true,
       },
       modDefinition,
     );
```
Comment (grahamlangford) [inline src/extensionConsole/pages/activateMod/ActivateModPage.test.tsx:215]:
```suggestion
    await waitFor(() => {
    expect(activateModCallbackMock).toHaveBeenCalledWith(
      {
        modComponents: { "0": true },
        optionsArgs: {},
        integrationDependencies: [],
        personalDeployment: true,
      },
      modDefinition,
    );
    });
```

**Review by grahamlangford — APPROVED** [review]

**Pull request merged**